### PR TITLE
[SPARK-14995][R] Add `since` tag in Roxygen documentation for SparkR API methods

### PR DESCRIPTION
**Pull request by dongjoon-hyun**
## What changes were proposed in this pull request?

This PR adds `since` tags to Roxygen documentation according to the previous documentation archive.

https://home.apache.org/~dongjoon/spark-2.0.0-docs/api/R/
## How was this patch tested?

Manual.

### CODE REVIEW

**Comment by dongjoon-hyun** [comment]
Hi, @shivaram , @felixcheung , @sun-rui .

It's the first draft. There is a little ambiguity like the following.

- `SparkDataFrame` is marked as `@note since 2.0.0` because `DataFrame` changed into `SparkDataFrame` in Spark 2.0.
- However, I used the oldest versions for the functions using `SparkDataFrame`. (They used `DataFrame` before.)

I think this is more natural.

File: R/pkg/R/SQLContext.R [hunk 1]
```diff
@@ -213,7 +213,7 @@ createDataFrame <- function(x, ...) {
 #' @aliases createDataFrame
 #' @export
 #' @method as.DataFrame default
-
+#' @note since 2.0.0
```
Comment (felixcheung) [inline R/pkg/R/SQLContext.R]: I thought we had as.DataFrame before 2.0?

File: R/pkg/R/generics.R [hunk 2]
```diff
@@ -20,157 +20,196 @@
 # @rdname aggregateRDD
 # @seealso reduce
 # @export
+# @note since 1.5.0
```
Comment (felixcheung) [inline R/pkg/R/generics.R]: I'm not sure we need to put @note on the generic. Basically this merges with the other roxygen block because of the @rdname tag, so it should already has one that. More concerning is whether doing it here would double it in the output html file - could you check?

File: R/pkg/R/SQLContext.R (hunk 1 above)
Comment (dongjoon-hyun) [inline R/pkg/R/SQLContext.R]: Oh, right. It must be 1.6.0. I'll fix that.

**Comment by felixcheung** [comment]
Cool! Thank you for pulling this together. Yea, I'm not sure how to handle SparkDataFrame either and I think what you have is reasonable. (SparkDataFrame should be fairly transparent to users anyway)

**Comment by felixcheung** [comment]
I reviewed all of it and they looked roughly correct for me, drawing from memory :) 👍

File: R/pkg/R/generics.R (hunk 2 above)
Comment (dongjoon-hyun) [inline R/pkg/R/generics.R]: Yep. I'll remove all notes on generics.R and check the output html.

**Comment by dongjoon-hyun** [comment]
Thank you for review, @felixcheung !

**Comment by dongjoon-hyun** [comment]
Hi, All.
For review, I uploaded the generated R doc here.

https://home.apache.org/~dongjoon/spark-2.0.0-docs/api/R/

The remaining issue is the **multiple** notes like the followings due to merging notes.

https://home.apache.org/~dongjoon/spark-2.0.0-docs/api/R/read.df.html

```
Note
since 1.4.0
since 1.6.0
```

How can we handle this nicely?

**Comment by felixcheung** [comment]
Hmm.. there isn't really a good way to do that due to how roxgen2 collapses the 2 blocks. In some cases other parts like parameter list could get very confusing too.

I think we could review them and consider separating them into different rd.
In this case it makes sense to have one rd for this since they are "alias", then perhaps the only way is to add to the note to say `#' @note read.df since 1.6.0"

**Comment by shivaram** [comment]
Yeah since the notes appear in a separate section I think its better to be more explicit -- so `read.df since 1.6.0` will be good for these cases

**Comment by felixcheung** [comment]
hopefully we don't have a lot of these...

**Comment by dongjoon-hyun** [comment]
Thank you for opinions! I'll revise and update the html doc.

**Comment by dongjoon-hyun** [comment]
Hi, @shivaram and @felixcheung .
Now the document is updated with master and shows merged notes correctly. I manually checked all the merged notes and used function signatures for overloaded functions.
```
Note
sparkR.stop since 1.4.0
sparkR.session.stop since 2.0.0
```
https://home.apache.org/~dongjoon/spark-2.0.0-docs/api/R/sparkR.session.stop.html

File: R/pkg/R/DataFrame.R [hunk 3]
```diff
@@ -59,6 +60,7 @@ setMethod("initialize", "SparkDataFrame", function(.Object, sdf, isCached) {
 #' @export
 #' @param sdf A Java object reference to the backing Scala DataFrame
 #' @param isCached TRUE if the SparkDataFrame is cached
+#' @note dataFrame since 2.0.0
```
Comment (felixcheung) [inline R/pkg/R/DataFrame.R]: you might not need to call this out with a separate note btw, this is an internal function to help wrap the returned Java object "handle" to a R class, not intend to be called by users.

**Comment by dongjoon-hyun** [comment]
Thank you, @felixcheung . I removed the note on `dataFrame`.

**Comment by dongjoon-hyun** [comment]
Hi, @shivaram and @felixcheung .
Now, PR and the generated HTML site are up-to-date again. (For HTML site, you may need to refresh.)

**Comment by felixcheung** [comment]
Do you think we should annotate with name even when they are on the same rd/page?

eg. there's only approxCountDistinct on this
image
(btw, the reason for duplicated "see also" is from multiple @family tag - could you clean them up if you see them?)

Originally my thought was we would only add name in front of version when there were multiple function going to the same page (eg. agg, summarize) AND that they have different starting release.

Like
image

But not in this case
image
(Here's another bug, count shouldn't collapse with agg/summarize?)

**Comment by dongjoon-hyun** [comment]
It seems you're concerning multiple issue. I'll focus on the same versions issue first.

The principle of this PR is simply adding `since` tag for all exposed functions.

IMHO, that is better since we should consider how to maintain the versions in the future. 

For example, the case of `count`, `agg`, `summarize`, let us assume that we add another function `agg2` in ther future.
 - Case 1: If we didn't keep version 1.4.0 for all functions, new contributor should update old functions together.
 - Case 2: If we keep version 1.4.0 for all functions, new contributor can add `agg2` without considering old functions.

I think `Case 1` is not good for maintainability.

**Comment by dongjoon-hyun** [comment]
By the way, before discussing further, should I retarget this PR for Spark 2.1.0 ?
I think this PR misses the deadline a little bit.

**Comment by dongjoon-hyun** [comment]
For other important issue about `see also`, all the previous doc look like that.

http://spark.apache.org/docs/1.6.0/api/R/approxCountDistinct.html
http://spark.apache.org/docs/1.5.0/api/R/approxCountDistinct.html
http://spark.apache.org/docs/latest/api/R/approxCountDistinct.html

So, it's a little bit beyond of the scope of this PR. It's long lasting document restructuring issue.
I think we need to design(or decide) how to restructure R doc by considering `raname`, `family` and `see also` issue all together.

**Comment by felixcheung** [comment]
I'm not sure what is the process but historically doc only changes are ok to go a bit later in the RC/release cut off.

I am ok with `@note` style though I'm concerned that it would be hard to follow for casual R contributors.

Regarding the "see also" or the overall doc structure, yes, that has been an ongoing issue with R doc. Most of others are ok and there is a pattern to follow. It is outside of the scope of this PR. These are bugs in the doc.

**Comment by shivaram** [comment]
- Yeah we can make documentation changes after the RC as the doc updates are pushed separately to the Spark website. However I prefer to get R doc changes in before the RC as these are the ones that users will see when they do a `?dapply`. 

- Also I think the version numbers in `@note` is a bit of an experiment at this point -- longer term this is something we should discuss with the `roxygen2` authors and see whats a good way forward.

- Finally I agree that there are existing issues when we merge multiple functions in the same rd file. It will be good to tackle those, but we can use a different PR for that

**Comment by felixcheung** [comment]
so if we are ok with `@note approxCountDistinct(Column) since 1.4.0` as the form for now I think we are ok with this PR.

**Comment by dongjoon-hyun** [comment]
Thank you for narrowing the scope. Indeed, this seems to be an experimental attempt.
I rebased to resolved the conflict. As @felixcheung mentioned, the use of function signature is the remains.
The function signature comes due to the following: https://home.apache.org/~dongjoon/spark-2.0.0-docs/api/R/show.html
```
Note
show(Column) since 1.4.0
show(GroupedData) since 1.4.0
show(SparkDataFrame) since 1.4.0
show(WindowSpec) since 2.0.0
```

**Comment by shivaram** [comment]
Ok. I think this is a reasonable proposal. I will take one more final pass on this PR today

**Comment by dongjoon-hyun** [comment]
Thank you, @shivaram !

**Comment by shivaram** [comment]
LGTM. Thanks for this PR @dongjoon-hyun -- This is very useful to have going forward.

Merging this to master, branch-2.0

**Comment by dongjoon-hyun** [comment]
Thank you, @shivaram and @felixcheung .
I'm really happy since this became a part of Spark 2.0 !!
